File: samples/AndroidSamples/salmon-android-samples/src/main/java/com/mku/salmon/samples/main/FileActivity.java
package com.mku.salmon.samples.main;

import android.content.Intent;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;

import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.textfield.TextInputEditText;
import com.mku.android.fs.file.AndroidFileSystem;
import com.mku.fs.file.IFile;
import com.mku.salmon.Generator;
import com.mku.salmon.samples.R;
import com.mku.salmon.samples.samples.FileSample;
import com.mku.salmon.samples.samples.SamplesCommon;
import com.mku.salmon.samples.utils.AndroidFileChooser;
import com.mku.salmon.streams.ProviderType;
import com.mku.salmon.streams.AesStream;

import java.io.IOException;


public class FileActivity extends AppCompatActivity {
    private static final int REQUEST_SAVE_FILE = 1000;
    private static final int REQUEST_LOAD_FILE = 1001;
    private TextInputEditText password;
    private EditText plainText;
    private EditText decryptedText;
    private EditText outputText;
    private Button saveButton;
    private Button loadButton;

    private static final String defaultPassword = "test123";
    private String text = "This is a plain text that will be encrypted";

    private byte[] key;
    byte[] integrityKey = null;


    @Override
    public void onCreate(Bundle bundle) {
        super.onCreate(bundle);

        setContentView(R.layout.activity_file);

        password = findViewById(R.id.TEXT_PASSWORD);

        saveButton = findViewById(R.id.SAVE_BUTTON);
        saveButton.setOnClickListener((e) -> {
            AndroidFileChooser.openFilesystem(this, REQUEST_SAVE_FILE, true);
        });

        loadButton = findViewById(R.id.LOAD_BUTTON);
        loadButton.setOnClickListener((e) -> {
            AndroidFileChooser.openFilesystem(this, REQUEST_LOAD_FILE, false);
        });

        plainText = findViewById(R.id.PLAIN_TEXT);
        plainText.setText(text);
        decryptedText = findViewById(R.id.DECRYPTED_TEXT);
        outputText = findViewById(R.id.OUTPUT_TEXT);
        password.setText(defaultPassword);

        initialize();
    }

    public void log(String msg) {
        runOnUiThread(() -> {
            outputText.append(msg + "\n");
        });
    }

    private void initialize() {
        AndroidFileSystem.initialize(this);
        AesStream.setAesProviderType(ProviderType.Default);
    }

    public void saveFile(IFile dir) {
        boolean integrity = true;

        outputText.setText("");
        // generate an encryption key from the text password
        key = SamplesCommon.getKeyFromPassword(password.getText().toString());

        // enable integrity (optional)
        byte[] integrityKey = null;
        if (integrity) {
            // generate an HMAC key
            integrityKey = Generator.getSecureRandomBytes(32);
        }

        String filename = "encrypted_data.dat";
        IFile file = dir.getChild(filename);
        if (file.exists()) {
            file.delete();
            file = dir.getChild(filename);
        }

        try {
            FileSample.encryptTextToFile(plainText.getText().toString(), key, integrityKey, file,
                    (msg) -> {
                        log(msg);
                    });
            log("file saved");
        } catch (Exception e) {
            e.printStackTrace();
            log(e.getMessage());
        }
    }

    public void loadFile(IFile file) {
        try {
            String decText = FileSample.decryptTextFromFile(key, integrityKey, file, (msg) -> {
                log(msg);
            });
            log("file loaded");
            decryptedText.setText(decText);
        } catch (Exception e) {
            e.printStackTrace();
            log(e.getMessage());
        }
    }


    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (data == null)
            return;
        android.net.Uri uri = data.getData();
        AndroidFileChooser.setUriPermissions(this, data, uri);
        switch (requestCode) {
            case REQUEST_SAVE_FILE:
                IFile saveDir = AndroidFileChooser.getFile(this, uri.toString(), true);
                saveFile(saveDir);
                break;
            case REQUEST_LOAD_FILE:
                IFile filesToImport = AndroidFileChooser.getFiles(this, data)[0];
                loadFile(filesToImport);
                break;
        }
    }
}

File: samples/DotNetAndroidSamples/DotNetAndroidSamples/Main/FileActivity.cs
using AndroidX.AppCompat.App;
using Android.Content;
using Android.Widget;
using Android.OS;
using System;
using DotNetAndroidSamples;
using Google.Android.Material.TextField;
using Mku.Android.FS.File;
using Mku.FS.File;
using Mku.Salmon.Samples.Samples;
using Mku.Salmon.Streams;
using Mku.Salmon.Samples.Utils;
using Android.App;
using Uri = Android.Net.Uri;
using System.Threading.Tasks;
using Mku.Salmon.Transform;
using Mku.Android.Salmon.Transform;

namespace Mku.Salmon.Samples.Main;

[Activity(Label = "@string/app_name", Theme = "@style/AppTheme",
    ScreenOrientation = global::Android.Content.PM.ScreenOrientation.Portrait)]
public class FileActivity : AppCompatActivity
{
    private const int REQUEST_SAVE_FILE = 1000;
    private const int REQUEST_LOAD_FILE = 1001;
    private TextInputEditText password;
    private EditText plainText;
    private EditText decryptedText;
    private EditText outputText;
    private Button saveButton;
    private Button loadButton;

    private static readonly string defaultPassword = "test123";
    private string text = "This is a plain text that will be encrypted";

    private byte[] key;
    byte[] integrityKey = null;


    protected override void OnCreate(Bundle bundle)
    {
        base.OnCreate(bundle);

        SetContentView(Resource.Layout.activity_file);

        password = (TextInputEditText)FindViewById(Resource.Id.TEXT_PASSWORD);

        saveButton = (Button)FindViewById(Resource.Id.SAVE_BUTTON);
        saveButton.Click += (s, e) =>
        {
            AndroidFileChooser.OpenFilesystem(this, REQUEST_SAVE_FILE, true);
        };

        loadButton = (Button)FindViewById(Resource.Id.LOAD_BUTTON);
        loadButton.Click += (s, e) =>
        {
            AndroidFileChooser.OpenFilesystem(this, REQUEST_LOAD_FILE, false);
        };

        plainText = (EditText)FindViewById(Resource.Id.PLAIN_TEXT);
        plainText.Text = text;
        decryptedText = (EditText)FindViewById(Resource.Id.DECRYPTED_TEXT);
        outputText = (EditText)FindViewById(Resource.Id.OUTPUT_TEXT);
        password.Text = defaultPassword;

        Initialize();
    }

    public void Log(string msg)
    {
        RunOnUiThread(() =>
        {
            outputText.Append(msg + "\n");
        });
    }

    public void ClearLog()
    {
        RunOnUiThread(() =>
        {
            outputText.Text = "";
        });
    }

    private void Initialize()
    {
		NativeTransformer.NativeProxy = new AndroidNativeProxy();
        AndroidFileSystem.Initialize(this);
        AesStream.AesProviderType = ProviderType.Default;
    }

    public void SaveFile(IFile dir)
    {
        bool integrity = true;

        ClearLog();
        // generate an encryption key from the text password
        key = SamplesCommon.GetKeyFromPassword(password.Text.ToString());

        // enable integrity (optional)
        byte[] integrityKey = null;
        if (integrity)
        {
            // generate an HMAC key
            integrityKey = Generator.GetSecureRandomBytes(32);
        }

        string filename = "encrypted_data.dat";
        IFile file = dir.GetChild(filename);
        if (file.Exists)
        {
            file.Delete();
            file = dir.GetChild(filename);
        }

        try
        {
            FileSample.EncryptTextToFile(plainText.Text.ToString(), key, integrityKey, file, Log);
            Log("file saved");
        }
        catch (Exception e)
        {
            Console.Error.WriteLine(e);
            Log(e.Message);
        }
    }

    public void LoadFile(IFile file)
    {
        try
        {
            string decText = FileSample.DecryptTextFromFile(key, integrityKey, file, Log);
            Log("file loaded");
            RunOnUiThread(() =>
            {
                decryptedText.Text = decText;
            });
        }
        catch (Exception e)
        {
            Console.Error.WriteLine(e);
            Log(e.Message);
        }
    }


    protected override void OnActivityResult(int requestCode, Result resultCode, Intent data)
    {
        base.OnActivityResult(requestCode, resultCode, data);
        if (data == null)
            return;
        Uri uri = data.Data;
        AndroidFileChooser.SetUriPermissions(this, data, uri);
        switch (requestCode)
        {
            case REQUEST_SAVE_FILE:
                IFile saveDir = AndroidFileSystem.GetRealFile(uri.ToString(), true);
                Task.Run(() =>
                {
                    SaveFile(saveDir);
                });
                break;
            case REQUEST_LOAD_FILE:
                IFile filesToImport = AndroidFileChooser.GetFiles(this, data)[0];
                Task.Run(() =>
                {
                    LoadFile(filesToImport);
                });
                break;
        }
    }
}